File: android_views/src/main/java/com/app/views/viewstate/TopicViewState.kt
package com.app.views.viewstate

import android.view.View
import com.app.shared.business.TopicState

sealed class TopicViewState(state: TopicState, noBookmarks: Int) {
    val id: Int = state.id
    val name: String = state.name
    val topicState: TopicState = state
    val isEditable: Boolean = state.isEditable
    val numberOfBookmarks = noBookmarks

    data class Normal(val state: TopicState, val noBookmarks: Int): TopicViewState(state, noBookmarks)

    data class Selectable(val state: TopicState, val isSelected: Boolean, val noBookmarks: Int): TopicViewState(state, noBookmarks) {
        val checkMarkVisibility = if (isSelected) View.VISIBLE else View.INVISIBLE
    }
}

File: ios_app/ios_views/ViewStates/TopicViewState.swift
//
//  TopicViewState.swift
//  ios_views
//
//  Created by Gabriel Coman on 05/05/2020.
//  Copyright © 2020 Gabriel Coman. All rights reserved.
//

import Foundation
import RememberShared

public protocol TopicViewStateType: Identifiable {
    var topic: TopicState { get }
    var noBookmarks: Int { get }
}

public extension TopicViewStateType {
    
    var id: Int32 {
        return topic.id
    }
    
    var name: String {
        return topic.name.uppercased()
    }
    
    var isEditable: Bool {
        return topic.isEditable
    }
}

public class TopicViewState {
    
    public struct Normal: TopicViewStateType {
        public let topic: TopicState
        public let noBookmarks: Int
         
        
        public init(state: TopicState, noBookmarks: Int) {
            self.topic = state
            self.noBookmarks = noBookmarks
        }
    }
    
    public struct Selectable: TopicViewStateType {
        public let topic: TopicState
        public let isSelected: Bool
        public let noBookmarks: Int
        
        public init(state: TopicState, isSelected: Bool, noBookmarks: Int) {
            self.topic = state
            self.isSelected = isSelected
            self.noBookmarks = noBookmarks
        }
    }
}
